### DISCUSSION TITLE
Hybrid Tenancy SaaS App Structure

### DISCUSSION THREAD

**Discussion by gdrose** [pocketbase/pocketbase#2999] [Q&A]
I'm currently structuring a SaaS project, and I wanted to follow the Hybrid Tenancy approach which consists in having a single webapp (in my case sveltekit) and then a database for each tenant. This would provide me with the data isolation that I'm required to have but a lot of questions popped up.

1) In order to follow this approach I would still need a central database for my app  to manage users, and then I would need to switch up the database connection after they are logged in to populate the software with their data.

2) I have to keep alive the connection with my database in my hooks in order to keep the authentication going for the user, which means that I'm going to have two pocketbase connections open at the same time, is this the right approach ? (I know that Dynamic Database connections are a thing, but I'm not sure if pocketbase is capable and/or should be used like this)

3) I loved pocketbase for a lot of client work that I did so I wanted to use it in this project too but I'm not sure if it's the best fit, any tips?

**Comment by ganigeorgiev** [comment]
I'm not entirely sure why you may need 2), but solely based on the above description I think PocketBase may not be suited for your use case and it may end up actually complicating things for you instead of making them easier.

**A single PocketBase process can work only with one SQLite database out of the box**. There is no builtin mechanism that will switch only the database files per request basis, if that is what you are trying to do.

If you already have a standalone server-side app (eg. the node adapter of SvelteKit) you may be better using only it and combine it with an ORM like Drizzle or Prisma to communicate directly with the databases of your clients (_in the end you don't have to use PocketBase to use SQLite_).

**Comment by gdrose** [comment]
I'm sorry if I created some confusion when explaining what I was trying to do. Basically In the second point I'm saying that I would be creating two pocketbase processes in my request handler (hooks.server.js), one for the user authentication in my own database and another one that connects to the tenant database URL (which is stored in my database). Basically every client would have their own pocketbase running on different machines and then I would have my pocketbase which contains all clients users. Is this approach suitable for pocketbase? 

I tried to test this approach but I got into some confusion because when doing so, I would have to authenticate both connections, meaning that probably the cookie would override itself (?); I know that's really confusing, sorry I'm trying my best to explain.

I also looked into Drizzle and Prisma but I don't really like their approach.

**Reply by ganigeorgiev** [reply]
> I tried to test this approach but I got into some confusion because when doing so, I would have to authenticate both connections, meaning that probably the cookie would override itself (?); 

I don't understand what do you mean by both connections. As far as I understand you have 1 "manager/supervisor" PocketBase instance that will contain information about the clients and where their individual PocketBase instances reside. The client itself will not be a user in their PocketBase instance, aka. each client will have their own set of PocketBase users, right? If that's the case, why do you have to maintain 2 connections?

How do you plan the client to interact with their PocketBase instance(s) in the first place?

**Reply by ganigeorgiev** [reply]
Also keep in mind that the JS SDK instance in the `hooks.server.js` uses the cookie only as "semi-persistent transport mechanism" to share the auth state between the client(browser) and the node SvelteKit app. The authentication in PocketBase doesn't work with cookies and it is enough to send valid `Authorization: TOKEN` header with a request to be considered authenticated (_this header is added automatically by the SDKs if there is loaded `pb.authStore.token`, hence why we have to load the cookie value in the `pb.authStore`_).

**Reply by gdrose** [reply]
After reading your feedback I got to this approach:

```
export const handle = async ({ event, resolve }) => {
	event.locals.pb = new PocketBase(MY_POCKETBASE_URL);
	event.locals.pb.authStore.loadFromCookie(event.request.headers.get('cookie') || '');

	try {
		if (event.locals.pb.authStore.isValid) {
			await event.locals.pb.collection('users').authRefresh();
			event.locals.user = serializeNonPOJOs(event.locals.pb.authStore.model);

			event.locals.tenant_pb = new PocketBase(event.locals.user.tenant_auth.ip);

			await event.locals.tenant_pb.admins.authWithPassword(
				event.locals.user.tenant_auth.email,
				event.locals.user.tenant_auth.password
			);
		}
	} catch (_) {
		event.locals.pb.authStore.clear();
		event.locals.user = undefined;
	}

	const response = await resolve(event);

	response.headers.set('set-cookie', event.locals.pb.authStore.exportToCookie({ secure: false }));

	return response;
};
```

From all the routes when I have to access the user's informations I just use locals.tenant_pb.  Is this semi-decent? Is it recommended to use this approach?

Thank you so much for the feedback and help so far.

**Reply by ganigeorgiev** [reply]
It seems that you are using the tenant email and password like Client ID and Client Secret.
This is OK but keep in mind that users usually don't tend to use complex passwords and may hinder your security.

Additionally calling `authWithPassword` on every request could have a negative impact on the performance of your app due to the bcrypt hash verification and you'll need some sort of cache to prevent calling it every time, eg. store the `authWithPassword` token as a different cookie in a similar manner to the  "users" one and call `authWithPassword` again only if it has expired, otherwise - `authRefresh`. To export/load a cookie with different name, you can pass a second argument to `exportToCookie(options, key)` and `loadFromCookie(cookie, key)`, something like:

```js
// to load a cookie with custom name:
event.locals.tenant_pb.authStore.loadFromCookie(cookie, "tenant_auth");

// to export a cookie with custom name:
const tenantCookie = event.locals.tenant_pb.authStore.exportToCookie({}, "tenant_auth");
```
